### DISCUSSION TITLE
Exception of type 'RabbitMQ.Client.Exceptions.PublishException' was thrown

### DISCUSSION THREAD

**Discussion by plewam** [rabbitmq/rabbitmq-dotnet-client#1701] [Q&A] [answered]
Hi All,

I am currently testing around with `7.0.0-rc.12`. However (I may be doing something wrong) I am no longer able to publish messages. I received a `PublishException` but do not get any further information with that. The exception does not contain any further info. Are you able to help?

I use RMQ Server 4.0.2 with Erlang 26.2.5.4

```
RabbitMQ.Client.Exceptions.PublishException: Exception of type 'RabbitMQ.Client.Exceptions.PublishException' was thrown.
   at RabbitMQ.Client.Impl.ChannelBase.BasicPublishAsync[TProperties](String exchange, String routingKey, Boolean mandatory, TProperties basicProperties, ReadOnlyMemory`1 body, CancellationToken cancellationToken)
```

**Comment by plewam** [comment, marked as answer]
Right... I found the issue. The below configuration causes the issue.

```
CreateChannelOptions channelOptions = new()
{
    PublisherConfirmationsEnabled = true,
    PublisherConfirmationTrackingEnabled = true
};
```

Changing it to makes it work again.

```
CreateChannelOptions channelOptions = new()
{
    PublisherConfirmationsEnabled = true,
    PublisherConfirmationTrackingEnabled = false
};
```

**Comment by lukebakken** [comment]
Hello! Thank you for the report. This is very strange, since the test suite uses this configuration all over the place:

```
    PublisherConfirmationsEnabled = true,
    PublisherConfirmationTrackingEnabled = true
```

Is the stack trace you posted the complete stack?

If you can share code to reproduce, that would be great. I'll look into this today.

**Comment by plewam** [comment]
Yes it sadly is. I even debugged it to gather more information.

**Reply by lukebakken** [reply]
Thanks, I'll get back on this today, hopefully. I really appreciate you testing the releases as they come out.

**Comment by lukebakken** [comment]
@plewam is there a chance that, in your environment, you're always getting Nacks from RabbitMQ? I realize this is unlikely.

Are you publishing with `mandatory: true`?

**Reply by plewam** [reply]
I do not receive Nacks, but mandatory is indeed set to true.

**Reply by plewam** [reply]
However by reading the docs I realise I do not need to set the mandatory flag for my use case…

**Reply by lukebakken** [reply]
OK, you will get `PublishException` when `mandatory: true` and a message can't be routed to any queues. Does that happen to be the case?

Like I said, the test suite enables confirms and confirm tracking all over the place so I think this particular issue has something to do with your environment...?

**Reply by plewam** [reply]
I will have it checked shortly, give me a few days.  If that is the case a publish exception is the right thing to do inside the library.

thanks for your help!

**Reply by lukebakken** [reply]
OK, thank you again for testing!

**Reply by plewam** [reply]
So I tested it again. Your library behaves correctly. I had the mandatory flag set and no queue bound to my destination exchange. Therefore it is correct to throw the `PublishException`. Thanks again for your help and assistance.

**Reply by lukebakken** [reply]
Thanks for the follow-up! I'm happy to assist, anytime.

**Comment by romerod** [comment]
Is this https://github.com/rabbitmq/rabbitmq-tutorials/blob/main/dotnet/PublisherConfirms/PublisherConfirms.cs example still ok?

It's not catching the PublishException which is thrown by BasicPublishAsync now, but relying on the BasicReturnAsync?

Is PublishException enough now to know that message was not routed?

**Reply by michaelklishin** [reply]
As [file history suggests](https://github.com/rabbitmq/rabbitmq-tutorials/commits/main/dotnet/PublisherConfirms/PublisherConfirms.cs), it was updated for the 7.x series of the client.

There's also [tutorial 7](https://www.rabbitmq.com/tutorials/tutorial-seven-dotnet) that focuses on publisher confirms specifically.

Publisher confirms are orthogonal to [unroutable message handling](https://www.rabbitmq.com/docs/publishers#unroutable).

**Comment by romerod** [comment]
Ah ok, thank you
The code https://github.com/rabbitmq/rabbitmq-tutorials/commit/a4adce94c9b326f7090d30b114eb9c62f8aba8b8 where PublishException was explicitly catched seemed to make more sense for me, but ok.

Thanks
